### PR TITLE
ARROW-14923: [Java] AllocationListener should be called during ownership transferring

### PR DESCRIPTION
**Pull request by zhztheplayer**

### CODE REVIEW

**Comment by pitrou** [comment]
@davisusanibar Would you like to take a look?

File: java/memory/memory-core/src/main/java/org/apache/arrow/memory/BufferLedger.java [hunk 1]
```diff
@@ -364,8 +364,12 @@ boolean transferBalance(final ReferenceManager targetReferenceManager) {
...
       }
 
-      boolean overlimit = targetAllocator.forceAllocate(allocationManager.getSize());
-      allocator.releaseBytes(allocationManager.getSize());
+      long size = allocationManager.getSize();
+      targetAllocator.getListener().onPreAllocation(size);
+      boolean overlimit = targetAllocator.forceAllocate(size);
+      targetAllocator.getListener().onAllocation(size);
+      allocator.releaseBytes(size);
+      allocator.getListener().onRelease(size);
```
Comment (davisusanibar) [inline java/memory/memory-core/src/main/java/org/apache/arrow/memory/BufferLedger.java:372]: If we do not validate "assertEquals(1, l1.getNumReleaseCalls())" or "assertEquals(1, l2.getNumCalls())", question: What could be the errors in current programs to this changes should be needed ?

Comment (zhztheplayer) [reply java/memory/memory-core/src/main/java/org/apache/arrow/memory/BufferLedger.java:372]: For example. Buffer1 and Buffer2 are shared and were created by Allocator1 and Allocator2. Allocator1 has a listener assigned but Allocator2 doesn't have one. Once the memory ownership is transferred from Buffer1 to Buffer2, then the listener will not be aware of that transferring, and `listener.free()` will never be called even we close both Buffer1 and Buffer2.

File: java/memory/memory-netty/src/test/java/org/apache/arrow/memory/TestBaseAllocator.java [hunk 2]
```diff
@@ -159,6 +159,87 @@ public void testAllocator_transferOwnership() throws Exception {
...
+
+      // do transfer
+      OwnershipTransferResult transferResult = buffer1.getReferenceManager()
+          .transferOwnership(buffer1, childAllocator2);
+      assertTrue(transferResult.getAllocationFit());
+      ArrowBuf buffer2 = transferResult.getTransferredBuffer();
+
+      // reservations from l1 should also be transferred to l2 during buffer transferring
+      assertEquals(1, l1.getNumCalls());
+      assertEquals(1, l1.getNumReleaseCalls());
```
[the remark this answers is not in the document]
Comment (zhztheplayer) [reply java/memory/memory-netty/src/test/java/org/apache/arrow/memory/TestBaseAllocator.java:189]: Umm... Do we have to consider the test case as a specific fix for this specific bug/issue?

[the remark this answers is not in the document]
Comment (zhztheplayer) [reply java/memory/memory-netty/src/test/java/org/apache/arrow/memory/TestBaseAllocator.java:189]: For example: line 188 can also be failed if someone made a regression change to let the `l1.onAllocate()` to be called again during transferring.

File: java/memory/memory-core/src/main/java/org/apache/arrow/memory/BufferLedger.java (hunk 1 above)
Comment (davisusanibar) [reply java/memory/memory-core/src/main/java/org/apache/arrow/memory/BufferLedger.java:372]: Ok thank you for clarification

File: java/memory/memory-netty/src/test/java/org/apache/arrow/memory/TestBaseAllocator.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (davisusanibar) [reply java/memory/memory-netty/src/test/java/org/apache/arrow/memory/TestBaseAllocator.java:189]: Ok thank you for clarification.

**Review by davisusanibar — APPROVED** [review]

**Comment by amol-** [comment]
Closing because it has been untouched for a while, in case it's still relevant feel free to reopen and move it forward 👍

**Pull request closed without merging**